File: src/vost/mirror.py
"""Mirror (backup/restore) operations for vost.

Ref-level mirroring: push all local refs to a remote (backup) or fetch
all remote refs to local (restore).  Extracted from repo.py and cli.py.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from dulwich.client import get_transport_and_path as _get_transport_and_path
from dulwich.errors import NotGitRepository
from dulwich.porcelain import ls_remote as _ls_remote
from dulwich.protocol import ZERO_SHA as _ZERO_SHA
from dulwich.repo import Repo as _DRepo

if TYPE_CHECKING:
    from .repo import GitStore


# ---------------------------------------------------------------------------
# Data structures
# ---------------------------------------------------------------------------

@dataclass
class RefChange:
    """A single ref change in a :class:`MirrorDiff`.

    Attributes:
        ref: Full ref name (e.g. ``"refs/heads/main"``).
        old_target: Previous 40-char hex SHA, or ``None`` for creates.
        new_target: New 40-char hex SHA, or ``None`` for deletes.
    """
    ref: str
    old_target: str | None = None
    new_target: str | None = None


@dataclass
class MirrorDiff:
    """Result of a :meth:`~vost.GitStore.backup` or :meth:`~vost.GitStore.restore` operation.

    Attributes:
        add: Refs to create.
        update: Refs to update.
        delete: Refs to delete.
    """
    add: list[RefChange] = field(default_factory=list)
    update: list[RefChange] = field(default_factory=list)
    delete: list[RefChange] = field(default_factory=list)

    @property
    def in_sync(self) -> bool:
        """``True`` if there are no changes."""
        return not self.add and not self.update and not self.delete

    @property
    def total(self) -> int:
        """Total number of ref changes."""
        return len(self.add) + len(self.update) + len(self.delete)


# ---------------------------------------------------------------------------
# Transport helpers (operate on raw dulwich Repo)
# ---------------------------------------------------------------------------

def _diff_refs(drepo: _DRepo, url: str, direction: str) -> dict:
    """Compare local and remote refs.

    *direction* is ``"push"`` (local->remote) or ``"pull"`` (remote->local).
    Returns ``{"create": [...], "update": [...], "delete": [...],
    "src": {ref: sha}, "dest": {ref: sha}}`` with bytes keys.
    """
    # Auto-create remote for push if it's a local path that doesn't exist
    is_local = not any(url.startswith(proto) for proto in ["http://", "https://", "git://", "ssh://"])
    if is_local and not url.startswith("file://"):
        # Detect scp-style URLs: user@host:path or host:path
        # Exclude Windows drive letters (single letter before colon).
        if "@" in url and ":" in url.split("@", 1)[1]:
            raise ValueError(
                f"scp-style URL not supported: {url!r} — use ssh:// format instead"
            )
        colon_idx = url.find(":")
        # A colon after >1 chars with no path separator before it
        # looks like host:path.  Treat both / and \ as separators
        # to avoid rejecting Windows paths (e.g. \\?\C:\repo).
        prefix = url[:colon_idx]
        if colon_idx > 1 and "/" not in prefix and "\\" not in prefix:
            raise ValueError(
                f"scp-style URL not supported: {url!r} — use ssh:// format instead"
            )
    if is_local and direction == "push":
        local_path = url[7:] if url.startswith("file://") else url
        if not os.path.exists(local_path):
            _DRepo.init_bare(local_path, mkdir=True)

    try:
        remote_result = _ls_remote(url)
        refs_dict = remote_result.refs if hasattr(remote_result, "refs") else remote_result
        remote_refs = {
            ref: sha
            for ref, sha in refs_dict.items()
            if ref != b"HEAD" and not ref.endswith(b"^{}")
        }
    except NotGitRepository:
        # Remote doesn't exist - treat as empty for push, fail for pull
        if direction == "push":
            remote_refs = {}
        else:
            raise

    local_refs = {
        ref: sha
        for ref, sha in drepo.get_refs().items()
        if ref != b"HEAD"
    }

    if direction == "push":
        src, dest = local_refs, remote_refs
    else:
        src, dest = remote_refs, local_refs

    create, update, delete = [], [], []
    for ref, sha in src.items():
        if ref not in dest:
            create.append(ref)
        elif dest[ref] != sha:
            update.append(ref)
    for ref in dest:
        if ref not in src:
            delete.append(ref)

    return {"create": create, "update": update, "delete": delete,
            "src": src, "dest": dest}


def _mirror_push(drepo: _DRepo, url: str, *, progress=None):
    """Push all local refs to *url*, mirroring (force + delete stale)."""
    client, path = _get_transport_and_path(url)
    local_refs = {
        ref: sha
        for ref, sha in drepo.get_refs().items()
        if ref != b"HEAD"
    }

    def update_refs(remote_refs):
        new_refs = {}
        for ref, sha in local_refs.items():
            new_refs[ref] = sha
        for ref in remote_refs:
            if ref not in local_refs and ref != b"HEAD":
                new_refs[ref] = _ZERO_SHA
        return new_refs

    def gen_pack(have, want, *, ofs_delta=False, progress=progress):
        return drepo.object_store.generate_pack_data(
            have, want, ofs_delta=ofs_delta, progress=progress,
        )

    return client.send_pack(path, update_refs, gen_pack, progress=progress)


def _mirror_fetch(drepo: _DRepo, url: str, *, progress=None):
    """Fetch all remote refs from *url*, mirroring (force + delete stale)."""
    client, path = _get_transport_and_path(url)
    result = client.fetch(path, drepo, progress=progress)

    remote_refs = {
        ref: sha
        for ref, sha in result.refs.items()
        if ref != b"HEAD" and not ref.endswith(b"^{}")
    }

    # Set all remote refs locally
    for ref, sha in remote_refs.items():
        drepo.refs[ref] = sha

    # Delete local refs not on remote
    for ref in list(drepo.refs.allkeys()):
        if ref != b"HEAD" and ref not in remote_refs:
            drepo.refs.remove_if_equals(ref, drepo.refs[ref])

    return result


# ---------------------------------------------------------------------------
# Core mirror functions
# ---------------------------------------------------------------------------

def _raw_diff_to_sync_diff(raw: dict) -> MirrorDiff:
    """Convert bytes-keyed diff dict to MirrorDiff."""
    src, dest = raw["src"], raw["dest"]

    def _sha(b):
        return b.decode() if isinstance(b, bytes) else str(b)

    add = [
        RefChange(ref=ref.decode(), new_target=_sha(src[ref]))
        for ref in raw["create"]
    ]
    update = [
        RefChange(ref=ref.decode(), old_target=_sha(dest[ref]), new_target=_sha(src[ref]))
        for ref in raw["update"]
    ]
    delete = [
        RefChange(ref=ref.decode(), old_target=_sha(dest[ref]))
        for ref in raw["delete"]
    ]
    return MirrorDiff(add=add, update=update, delete=delete)


def backup(store: GitStore, url: str, *, dry_run: bool = False, progress=None) -> MirrorDiff:
    """Push all refs to *url*, creating an exact mirror.

    Returns a `MirrorDiff` describing what changed (or would change).
    """
    drepo = store._repo._drepo
    raw = _diff_refs(drepo, url, "push")
    diff = _raw_diff_to_sync_diff(raw)
    if not dry_run:
        _mirror_push(drepo, url, progress=progress)
    return diff


def restore(store: GitStore, url: str, *, dry_run: bool = False, progress=None) -> MirrorDiff:
    """Fetch all refs from *url*, overwriting local state.

    Returns a `MirrorDiff` describing what changed (or would change).
    """
    drepo = store._repo._drepo
    raw = _diff_refs(drepo, url, "pull")
    diff = _raw_diff_to_sync_diff(raw)
    if not dry_run:
        _mirror_fetch(drepo, url, progress=progress)
    return diff


# ---------------------------------------------------------------------------
# Credentials
# ---------------------------------------------------------------------------

def resolve_credentials(url: str) -> str:
    """Inject credentials into an HTTPS URL if available.

    Tries ``git credential fill`` first (works with any configured helper:
    osxkeychain, wincred, libsecret, ``gh auth setup-git``, etc.).  Falls
    back to ``gh auth token`` for GitHub hosts.  Non-HTTPS URLs and URLs
    that already contain credentials are returned unchanged.
    """
    if not url.startswith("https://"):
        return url

    from urllib.parse import urlparse, urlunparse

    parsed = urlparse(url)
    if parsed.username:
        return url  # already has credentials

    import subprocess

    # Try git credential fill
    try:
        stdin = f"protocol={parsed.scheme}\nhost={parsed.hostname}\n\n"
        proc = subprocess.run(
            ["git", "credential", "fill"],
            input=stdin, capture_output=True, text=True, timeout=5,
        )
        if proc.returncode == 0:
            creds = {}
            for line in proc.stdout.strip().splitlines():
                if "=" in line:
                    k, _, v = line.partition("=")
                    creds[k] = v
            username = creds.get("username")
            password = creds.get("password")
            if username and password:
                from urllib.parse import quote

                netloc = f"{quote(username, safe='')}:{quote(password, safe='')}@{parsed.hostname}"
                if parsed.port:
                    netloc += f":{parsed.port}"
                return urlunparse(parsed._replace(netloc=netloc))
    except (FileNotFoundError, subprocess.TimeoutExpired):
        pass

    # Fallback: gh auth token (GitHub-specific)
    try:
        proc = subprocess.run(
            ["gh", "auth", "token", "--hostname", parsed.hostname],
            capture_output=True, text=True, timeout=5,
        )
        token = proc.stdout.strip()
        if proc.returncode == 0 and token:
            netloc = f"x-access-token:{token}@{parsed.hostname}"
            if parsed.port:
                netloc += f":{parsed.port}"
            return urlunparse(parsed._replace(netloc=netloc))
    except (FileNotFoundError, subprocess.TimeoutExpired):
        pass

    return url

File: kotlin/src/main/kotlin/vost/Mirror.kt
package vost

import org.eclipse.jgit.lib.Constants
import org.eclipse.jgit.lib.NullProgressMonitor
import org.eclipse.jgit.lib.ObjectId
import org.eclipse.jgit.lib.Repository
import org.eclipse.jgit.storage.file.FileRepositoryBuilder
import org.eclipse.jgit.transport.*
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.TimeUnit

// ---------------------------------------------------------------------------
// Credentials
// ---------------------------------------------------------------------------

/**
 * Percent-encode a string for use in URL userinfo.
 */
private fun percentEncode(s: String): String {
    val sb = StringBuilder()
    for (b in s.toByteArray(Charsets.UTF_8)) {
        val c = b.toInt() and 0xFF
        if (c in 'A'.code..'Z'.code || c in 'a'.code..'z'.code ||
            c in '0'.code..'9'.code || c == '-'.code || c == '_'.code ||
            c == '.'.code || c == '~'.code
        ) {
            sb.append(c.toChar())
        } else {
            sb.append(String.format("%%%02X", c))
        }
    }
    return sb.toString()
}

/**
 * Inject credentials into an HTTPS URL if available.
 *
 * Tries `git credential fill` first (works with any configured helper:
 * osxkeychain, wincred, libsecret, `gh auth setup-git`, etc.).  Falls
 * back to `gh auth token` for GitHub hosts.  Non-HTTPS URLs and URLs
 * that already contain credentials are returned unchanged.
 *
 * @param url The URL to resolve credentials for.
 * @return The URL with credentials injected, or the original URL.
 */
fun resolveCredentials(url: String): String {
    if (!url.startsWith("https://")) return url

    val afterScheme = url.substring(8) // after "https://"
    val pathStart = afterScheme.indexOf('/').let { if (it < 0) afterScheme.length else it }
    val authority = afterScheme.substring(0, pathStart)

    // Already has credentials
    if ("@" in authority) return url

    val host = authority // may include :port
    val hostname = host.split(":").first()
    val pathAndRest = afterScheme.substring(pathStart)

    // Try git credential fill
    try {
        val proc = ProcessBuilder("git", "credential", "fill")
            .redirectErrorStream(false)
            .start()
        proc.outputStream.write("protocol=https\nhost=$hostname\n\n".toByteArray())
        proc.outputStream.close()
        val output = proc.inputStream.bufferedReader().readText()
        if (proc.waitFor(5, TimeUnit.SECONDS) && proc.exitValue() == 0) {
            val creds = mutableMapOf<String, String>()
            for (line in output.trim().lines()) {
                val eq = line.indexOf('=')
                if (eq > 0) {
                    creds[line.substring(0, eq)] = line.substring(eq + 1).trim()
                }
            }
            val username = creds["username"]
            val password = creds["password"]
            if (username != null && password != null) {
                return "https://${percentEncode(username)}:${percentEncode(password)}@$host$pathAndRest"
            }
        } else {
            proc.destroyForcibly()
        }
    } catch (_: Exception) {
    }

    // Fallback: gh auth token (GitHub-specific)
    try {
        val proc = ProcessBuilder("gh", "auth", "token", "--hostname", hostname)
            .redirectErrorStream(false)
            .start()
        val token = proc.inputStream.bufferedReader().readText().trim()
        if (proc.waitFor(5, TimeUnit.SECONDS) && proc.exitValue() == 0 && token.isNotEmpty()) {
            return "https://x-access-token:$token@$host$pathAndRest"
        } else {
            proc.destroyForcibly()
        }
    } catch (_: Exception) {
    }

    return url
}

/**
 * Mirror (backup/restore) operations for vost.
 *
 * Ref-level mirroring: push all local refs to a remote (backup) or fetch
 * all remote refs to local (restore).
 */
internal object MirrorOps {

    /**
     * Push all local refs to url, creating an exact mirror.
     *
     * Without [refs] this is a full mirror: remote-only refs are deleted.
     * With [refs] only the specified refs are pushed (no deletes).
     *
     * @param store The GitStore to back up.
     * @param url Destination URL (local path or remote URL), or bundle file path.
     * @param dryRun If true, compute diff but don't push.
     * @param refs Optional list of ref names to limit the backup to.
     * @param format Optional format string; "bundle" forces bundle output.
     * @return MirrorDiff describing what changed (or would change).
     */
    fun backup(
        store: GitStore,
        url: String,
        dryRun: Boolean = false,
        refs: List<String>? = null,
        format: String? = null,
    ): MirrorDiff {
        val repo = store.repo
        val useBundle = format == "bundle" || isBundlePath(url)

        if (useBundle) {
            val diff = diffBundleExport(repo, refs)
            if (!dryRun) {
                bundleExport(repo, url, refs)
            }
            return diff
        }

        // Auto-create bare repo for local push targets
        autoCreateBareRepo(url)

        if (refs != null) {
            val localRefs = getLocalRefs(repo)
            val remoteRefs = getRemoteRefs(repo, url)
            val refSet = resolveRefNames(refs, localRefs.keys)
            val fullDiff = diffRefs(localRefs, remoteRefs)
            // Filter to only targeted refs, no deletes
            val diff = MirrorDiff(
                add = fullDiff.add.filter { it.refName in refSet },
                update = fullDiff.update.filter { it.refName in refSet },
                delete = emptyList(),
            )
            if (!dryRun && !diff.inSync) {
                targetedPush(repo, url, localRefs, refSet)
            }
            return diff
        }

        val localRefs = getLocalRefs(repo)
        val remoteRefs = getRemoteRefs(repo, url)
        val diff = diffRefs(localRefs, remoteRefs)

        if (!dryRun && !diff.inSync) {
            mirrorPush(repo, url, localRefs, remoteRefs)
        }

        return diff
    }

    /**
     * Fetch refs from url additively (no deletes).
     *
     * Restore is **additive**: it adds and updates refs but never deletes
     * local-only refs.
     *
     * @param store The GitStore to restore into.
     * @param url Source URL (local path or remote), or bundle file path.
     * @param dryRun If true, compute diff but don't fetch.
     * @param refs Optional list of ref names to limit the restore to.
     * @param format Optional format string; "bundle" forces bundle input.
     * @return MirrorDiff describing what changed (or would change).
     */
    fun restore(
        store: GitStore,
        url: String,
        dryRun: Boolean = false,
        refs: List<String>? = null,
        format: String? = null,
    ): MirrorDiff {
        val repo = store.repo
        val useBundle = format == "bundle" || isBundlePath(url)

        if (useBundle) {
            val diff = diffBundleImport(repo, url, refs)
            if (!dryRun) {
                bundleImport(repo, url, refs)
            }
            return diff
        }

        val localRefs = getLocalRefs(repo)
        val remoteRefs = getRemoteRefs(repo, url)
        // For restore, remote is source, local is destination
        var diff = diffRefs(remoteRefs, localRefs)

        if (refs != null) {
            val refSet = resolveRefNames(refs, remoteRefs.keys)
            diff = MirrorDiff(
                add = diff.add.filter { it.refName in refSet },
                update = diff.update.filter { it.refName in refSet },
                delete = emptyList(), // additive: never delete
            )
        } else {
            diff = diff.copy(delete = emptyList()) // additive: never delete
        }

        if (!dryRun && !diff.inSync) {
            additiveFetch(repo, url, remoteRefs, refs)
        }

        return diff
    }

    // ── Internal helpers ─────────────────────────────────────────────

    /**
     * Get all local refs (excluding HEAD) as {refName: sha_hex}.
     */
    private fun getLocalRefs(repo: Repository): Map<String, String> {
        val result = mutableMapOf<String, String>()
        for (ref in repo.refDatabase.refs) {
            val name = ref.name
            if (name == Constants.HEAD) continue
            val peeled = repo.refDatabase.peel(ref)
            val oid = peeled.peeledObjectId ?: peeled.objectId ?: continue
            result[name] = oid.name
        }
        return result
    }

    /**
     * Get all remote refs (excluding HEAD and ^{} markers) as {refName: sha_hex}.
     */
    private fun getRemoteRefs(repo: Repository, url: String): Map<String, String> {
        val result = mutableMapOf<String, String>()
        try {
            val transport = Transport.open(repo, URIish(url))
            try {
                val connection = transport.openFetch()
                try {
                    val refs = connection.refsMap
                    for ((name, ref) in refs) {
                        if (name == Constants.HEAD) continue
                        if (name.endsWith("^{}")) continue
                        val oid = ref.objectId ?: continue
                        result[name] = oid.name
                    }
                } finally {
                    connection.close()
                }
            } finally {
                transport.close()
            }
        } catch (_: Exception) {
            // Remote doesn't exist or is inaccessible — treat as empty
        }
        return result
    }

    /**
     * Compare source and destination refs, producing a MirrorDiff.
     *
     * @param src The source refs (what we want the dest to look like).
     * @param dest The destination refs (current state).
     */
    private fun diffRefs(src: Map<String, String>, dest: Map<String, String>): MirrorDiff {
        val add = mutableListOf<RefChange>()
        val update = mutableListOf<RefChange>()
        val delete = mutableListOf<RefChange>()

        for ((ref, sha) in src) {
            if (ref !in dest) {
                add.add(RefChange(refName = ref, oldTarget = null, newTarget = sha))
            } else if (dest[ref] != sha) {
                update.add(RefChange(refName = ref, oldTarget = dest[ref], newTarget = sha))
            }
        }

        for ((ref, sha) in dest) {
            if (ref !in src) {
                delete.add(RefChange(refName = ref, oldTarget = sha, newTarget = null))
            }
        }

        return MirrorDiff(add = add, update = update, delete = delete)
    }

    /**
     * Push all local refs to remote, force-updating and deleting stale refs.
     */
    private fun mirrorPush(
        repo: Repository,
        url: String,
        localRefs: Map<String, String>,
        remoteRefs: Map<String, String>,
    ) {
        val transport = Transport.open(repo, URIish(url))
        try {
            val commands = mutableListOf<RemoteRefUpdate>()

            // Push all local refs (force)
            for ((refName, sha) in localRefs) {
                val oid = ObjectId.fromString(sha)
                commands.add(
                    RemoteRefUpdate(
                        repo,
                        null as String?,  // source ref name
                        oid,
                        refName,          // remote ref name
                        true,             // force update
                        null,             // tracking ref
                        null,             // expected old object id
                    )
                )
            }

            // Delete stale remote refs
            for (refName in remoteRefs.keys) {
                if (refName !in localRefs) {
                    commands.add(
                        RemoteRefUpdate(
                            repo,
                            null as String?,
                            ObjectId.zeroId(),
                            refName,
                            true,
                            null,
                            null,
                        )
                    )
                }
            }

            if (commands.isNotEmpty()) {
                transport.push(NullProgressMonitor.INSTANCE, commands)
            }
        } finally {
            transport.close()
        }
    }

    /**
     * Fetch all remote refs, overwriting local state (destructive — deletes stale).
     */
    private fun mirrorFetch(
        repo: Repository,
        url: String,
        remoteRefs: Map<String, String>,
        localRefs: Map<String, String>,
    ) {
        // Build fetch ref specs for all remote refs
        val refSpecs = remoteRefs.keys.map { refName ->
            RefSpec("+$refName:$refName")
        }

        if (refSpecs.isNotEmpty()) {
            val transport = Transport.open(repo, URIish(url))
            try {
                transport.fetch(NullProgressMonitor.INSTANCE, refSpecs)
            } finally {
                transport.close()
            }
        }

        // Delete local refs that are not on remote
        for (refName in localRefs.keys) {
            if (refName !in remoteRefs) {
                val refUpdate = repo.updateRef(refName)
                refUpdate.isForceUpdate = true
                refUpdate.delete()
            }
        }
    }

    /**
     * Push only refs in [refFilter] to url (no deletes).
     */
    private fun targetedPush(
        repo: Repository,
        url: String,
        localRefs: Map<String, String>,
        refFilter: Set<String>,
    ) {
        val transport = Transport.open(repo, URIish(url))
        try {
            val commands = mutableListOf<RemoteRefUpdate>()
            for (refName in refFilter) {
                val sha = localRefs[refName] ?: continue
                val oid = ObjectId.fromString(sha)
                commands.add(
                    RemoteRefUpdate(
                        repo,
                        null as String?,
                        oid,
                        refName,
                        true,
                        null,
                        null,
                    )
                )
            }
            if (commands.isNotEmpty()) {
                transport.push(NullProgressMonitor.INSTANCE, commands)
            }
        } finally {
            transport.close()
        }
    }

    /**
     * Fetch refs from url additively (no deletes).
     *
     * If [refs] is given, only those refs are fetched.
     */
    private fun additiveFetch(
        repo: Repository,
        url: String,
        remoteRefs: Map<String, String>,
        refs: List<String>?,
    ) {
        val refsToFetch: Set<String> = if (refs != null) {
            resolveRefNames(refs, remoteRefs.keys)
        } else {
            remoteRefs.keys.toSet()
        }

        val refSpecs = refsToFetch.filter { it in remoteRefs }.map { RefSpec("+$it:$it") }

        if (refSpecs.isNotEmpty()) {
            val transport = Transport.open(repo, URIish(url))
            try {
                transport.fetch(NullProgressMonitor.INSTANCE, refSpecs)
            } finally {
                transport.close()
            }
        }
        // No deletes — additive
    }

    // ── Bundle helpers ──────────────────────────────────────────────────

    /**
     * Return true if [path] has a `.bundle` extension.
     */
    private fun isBundlePath(path: String): Boolean =
        path.lowercase().endsWith(".bundle")

    /**
     * Resolve short ref names to full ref paths.
     *
     * Tries `refs/heads/`, `refs/tags/`, `refs/notes/` prefixes against
     * [available].  Full paths (starting with `refs/`) pass through
     * unchanged.  If no match is found the name is assumed to be a branch.
     */
    private fun resolveRefNames(names: List<String>, available: Set<String>): Set<String> {
        val result = mutableSetOf<String>()
        for (name in names) {
            if (name.startsWith("refs/")) {
                result.add(name)
                continue
            }
            var found = false
            for (prefix in listOf("refs/heads/", "refs/tags/", "refs/notes/")) {
                val candidate = "$prefix$name"
                if (candidate in available) {
                    result.add(candidate)
                    found = true
                    break
                }
            }
            if (!found) {
                result.add("refs/heads/$name")
            }
        }
        return result
    }

    /**
     * Create a bundle file from local refs.
     */
    private fun bundleExport(repo: Repository, path: String, refs: List<String>?) {
        val allRefs = getLocalRefs(repo)
        val refsToExport: Set<String> = if (refs != null) {
            resolveRefNames(refs, allRefs.keys)
        } else {
            allRefs.keys.toSet()
        }

        val writer = BundleWriter(repo)
        for (refName in refsToExport) {
            val sha = allRefs[refName] ?: continue
            writer.include(refName, ObjectId.fromString(sha))
        }
        FileOutputStream(path).use { fos ->
            writer.writeBundle(NullProgressMonitor.INSTANCE, fos)
        }
    }

    /**
     * Import refs from a bundle file (additive — no deletes).
     */
    private fun bundleImport(repo: Repository, path: String, refs: List<String>?) {
        val bundleRefs = bundleListHeads(repo, path)
        val refsToImport: Map<String, String> = if (refs != null) {
            val resolved = resolveRefNames(refs, bundleRefs.keys)
            bundleRefs.filterKeys { it in resolved }
        } else {
            bundleRefs
        }

        if (refsToImport.isEmpty()) return

        val refSpecs = refsToImport.keys.map { RefSpec("+$it:$it") }
        val uri = URIish(File(path).toURI().toString())
        val transport = Transport.open(repo, uri)
        try {
            transport.fetch(NullProgressMonitor.INSTANCE, refSpecs)
        } finally {
            transport.close()
        }
    }

    /**
     * List refs in a bundle file.
     */
    private fun bundleListHeads(repo: Repository, path: String): Map<String, String> {
        val uri = URIish(File(path).toURI().toString())
        val transport = Transport.open(repo, uri)
        val result = mutableMapOf<String, String>()
        try {
            val connection = transport.openFetch()
            try {
                for ((name, ref) in connection.refsMap) {
                    if (name == Constants.HEAD) continue
                    if (name.endsWith("^{}")) continue
                    val oid = ref.objectId ?: continue
                    result[name] = oid.name
                }
            } finally {
                connection.close()
            }
        } finally {
            transport.close()
        }
        return result
    }

    /**
     * Compute diff for exporting a bundle (all refs are 'add').
     */
    private fun diffBundleExport(repo: Repository, refs: List<String>?): MirrorDiff {
        val localRefs = getLocalRefs(repo)
        val filtered = if (refs != null) {
            val resolved = resolveRefNames(refs, localRefs.keys)
            localRefs.filterKeys { it in resolved }
        } else {
            localRefs
        }
        return MirrorDiff(
            add = filtered.map { (refName, sha) ->
                RefChange(refName = refName, oldTarget = null, newTarget = sha)
            },
        )
    }

    /**
     * Compute diff for importing a bundle (additive — no deletes).
     */
    private fun diffBundleImport(repo: Repository, path: String, refs: List<String>?): MirrorDiff {
        val bundleRefs = bundleListHeads(repo, path)
        val filtered = if (refs != null) {
            val resolved = resolveRefNames(refs, bundleRefs.keys)
            bundleRefs.filterKeys { it in resolved }
        } else {
            bundleRefs
        }
        val localRefs = getLocalRefs(repo)
        val diff = diffRefs(filtered, localRefs)
        // Additive: no deletes
        return diff.copy(delete = emptyList())
    }

    /**
     * Auto-create a bare git repo at a local path if it doesn't exist.
     */
    private fun autoCreateBareRepo(url: String) {
        // Only for local paths (not remote URLs)
        if (url.startsWith("http://") || url.startsWith("https://") ||
            url.startsWith("git://") || url.startsWith("ssh://")) {
            return
        }

        val path = if (url.startsWith("file://")) url.removePrefix("file://") else url

        // Detect scp-style URLs
        if ("@" in url && ":" in url.split("@", limit = 2)[1]) return
        val colonIdx = url.indexOf(':')
        if (colonIdx > 1) {
            val prefix = url.substring(0, colonIdx)
            if ('/' !in prefix && '\\' !in prefix) return
        }

        val dir = File(path)
        if (dir.exists()) return

        // Create bare repo
        dir.mkdirs()
        val repo = FileRepositoryBuilder()
            .setBare()
            .setGitDir(dir)
            .build()
        repo.create(true)
        repo.close()
    }
}
